csi/moac/watcher.js

```diff
@@ -201,6 +201,51 @@ class Watcher extends EventEmitter {
     }
   }
 
+  // Fetches the latest object(s) from k8s and updates the cache; then it
+  // returns the k8s object(s) from the cache or null if it does not exist.
+  async getRawBypass (name) {
+    try {
+      const getObj = await this.getEp(name).get();
+      if (getObj.statusCode === 200) {
+        const k8sObj = getObj.body;
+        const cachedObj = this.objects[name];
+
+        if (!cachedObj) {
+          // we still haven't processed the "ADDED" event so add it now
+          this._processEvent({
+            type: 'ADDED',
+            object: k8sObj
+          });
+        } else if (!k8sObj.metadata.generation || cachedObj.metadata.generation < k8sObj.metadata.generation) {
+          // the object already exists so modify it
+          this._processEvent({
+            type: 'MODIFIED',
+            object: k8sObj
+          });
+        }
+      } else {
+        const code = getObj.statusCode;
+        log.error(`Failed to fetch latest "${name}" from k8s, code: "${code}". Will only use the cached values instead.`);
+      }
+    } catch (err) {
+      // an exception is thrown if the resource does not exist
+      if (err.code === 404) {
+        const cachedObj = this.objects[name];
+        if (cachedObj) {
+          // the object does not exist so delete it
+          this._processEvent({
+            type: 'DELETED',
+            object: cachedObj
+          });
```

Comment:
I would not rather handle 404 case and would return whatever we have in the cache instead. The reason is that if we delete the object from cache, the MODIFIED and DELETED events could still come in at some point. What happens when watcher sees MODIFIED and there is no object in the cache? It will create it and emit event that there is a new object (even though it does not exist at that point in time - confusing all listeners. More safe seems to be to return stale copy of the object from the cache and let the update operation done by the caller simply fail. It will also simplify the code in this function.

Reply:
I was trying to "short-circuit" the events, but I see your point, let's let the events settle on their own on this one.